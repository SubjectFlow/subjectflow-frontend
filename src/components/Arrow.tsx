import React from "react";
import "../styles/Arrow.css";
import { Point } from "../Types";

const xCurveFactor = 0.3;
const yCurveFactor = 0.06;
const curveFactor1 = 2;
const curveFactor2 = 2;
const headRadius = 8;

type ArrowProps = {
  start: Point;
  end: Point;
};

function Arrow(props: ArrowProps) {
  let bottom, left;

  const diff: Point = {
    x: Math.abs(props.end.x - props.start.x),
    y: Math.abs(props.end.y - props.start.y)
  };

  let svgLineStart: Point = {
    x: 0,
    y: 0
  };

  let svgLineEnd: Point = {
    x: diff.x,
    y: diff.y
  };

  if (props.start.x > props.end.x) {
    svgLineStart.x = diff.x;
    svgLineEnd.x = 0;
    left = props.end.x;
  } else {
    left = props.start.x;
  }

  if (props.start.y <= props.end.y) {
    svgLineStart.y = diff.y + headRadius;
    svgLineEnd.y = headRadius;
    bottom = props.start.y;
  } else bottom = props.end.y;

  const svgLineDiff: Point = {
    x: svgLineEnd.x - svgLineStart.x,
    y: svgLineEnd.y - svgLineStart.y
  };

  return (
    <svg
      className="arrow"
      width={diff.x}
      height={diff.y + headRadius}
      xmlns="http://www.w3.org/2000/svg"
      style={{ bottom: bottom, left: left }}
    >
      <path
        className="arrow__line"
        d={
          `M ${svgLineStart.x} ${svgLineStart.y} ` +
          `C ${svgLineStart.x + curveFactor1 * xCurveFactor * svgLineDiff.x} ` +
          `${svgLineStart.y + curveFactor1 * yCurveFactor * svgLineDiff.y}, ` +
          `${svgLineEnd.x - curveFactor2 * xCurveFactor * svgLineDiff.x} ` +
          `${svgLineEnd.y - curveFactor2 * yCurveFactor * svgLineDiff.y}, ` +
          `${svgLineEnd.x} ${svgLineEnd.y}`
        }
      />
      <circle
        className="arrow__head"
        cx={svgLineEnd.x}
        cy={svgLineEnd.y}
        r={headRadius}
      />
    </svg>
  );
}

export default Arrow;
