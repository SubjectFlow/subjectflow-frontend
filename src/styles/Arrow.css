.arrow {
  position: absolute;
}

.arrow__line {
  stroke: black;
  stroke-width: 1;
  fill: transparent;
}

.arrow__head {
  fill: black;
}
