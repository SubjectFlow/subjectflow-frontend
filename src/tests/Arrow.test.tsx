import React from "react";
import renderer from "react-test-renderer";
import Arrow from "../components/Arrow";

const bottomLeft = { x: 0, y: 0 };
const topRight = { x: 400, y: 400 };
const topLeft = { x: 0, y: 400 };
const bottomRight = { x: 400, y: 100 };

test("against arrow snapshots", () => {
  let component = renderer.create(
    <Arrow faded={false} start={bottomLeft} end={topRight} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component = renderer.create(
    <Arrow faded={false} start={topLeft} end={bottomRight} />
  );

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component = renderer.create(<Arrow faded={false} start={topLeft} end={bottomLeft} />);

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component = renderer.create(<Arrow faded={false} start={topRight} end={topLeft} />);

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
