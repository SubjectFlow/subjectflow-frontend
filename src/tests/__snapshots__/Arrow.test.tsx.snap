// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`against arrow snapshots 1`] = `
<svg
  className="arrow"
  height={408}
  style={
    Object {
      "bottom": 0,
      "left": 0,
    }
  }
  width={400}
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    className="arrow__line"
    d="M 0 408 C 240 360, 160 56, 400 8"
  />
  <circle
    className="arrow__head"
    cx={400}
    cy={8}
    r={8}
  />
</svg>
`;

exports[`against arrow snapshots 2`] = `
<svg
  className="arrow"
  height={308}
  style={
    Object {
      "bottom": 92,
      "left": 0,
    }
  }
  width={400}
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    className="arrow__line"
    d="M 0 0 C 240 36, 160 264, 400 300"
  />
  <circle
    className="arrow__head"
    cx={400}
    cy={300}
    r={8}
  />
</svg>
`;

exports[`against arrow snapshots 3`] = `
<svg
  className="arrow"
  height={408}
  style={
    Object {
      "bottom": -8,
      "left": 0,
    }
  }
  width={0}
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    className="arrow__line"
    d="M 0 0 C 0 48, 0 352, 0 400"
  />
  <circle
    className="arrow__head"
    cx={0}
    cy={400}
    r={8}
  />
</svg>
`;

exports[`against arrow snapshots 4`] = `
<svg
  className="arrow"
  height={8}
  style={
    Object {
      "bottom": 400,
      "left": 0,
    }
  }
  width={400}
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    className="arrow__line"
    d="M 400 8 C 160 8, 240 8, 0 8"
  />
  <circle
    className="arrow__head"
    cx={0}
    cy={8}
    r={8}
  />
</svg>
`;
